agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -16,58 +16,60 @@ class ServicePodWorker
     def initialize(node, service_pod)
       @node = node
       @service_pod = service_pod
-      @prev_state = 'unknown'
+      @prev_state = nil # sync'd to master
     end
 
     # @param [Kontena::Models::ServicePod] service_pod
     def update(service_pod)
-      if @service_pod && @service_pod.deploy_rev != service_pod.deploy_rev
-        @prev_state = nil
-      end
       @service_pod = service_pod
-      ensure_desired_state
+      apply
     end
 
     def destroy
       @service_pod.mark_as_terminated
-      ensure_desired_state
+      apply
     end
 
-    def ensure_desired_state
+    def apply
       exclusive {
-        debug "state of #{service_pod.name}: #{service_pod.desired_state}"
-        service_container = get_container(service_pod.service_id, service_pod.instance_number)
-        if service_pod.running? && service_container.nil?
-          info "creating #{service_pod.name}"
-          ensure_running
-        elsif service_container && service_pod.running? && !service_container.running?
-          info "starting #{service_pod.name}"
-          ensure_started
-        elsif service_pod.running? && (service_container && service_container_outdated?(service_container, service_pod))
-          info "re-creating #{service_pod.name}"
-          ensure_running
-        elsif service_pod.stopped? && (service_container && service_container.running?)
-          info "stopping #{service_pod.name}"
-          ensure_stopped
-        elsif service_pod.terminated?
-          info "terminating #{service_pod.name}"
-          ensure_terminated if service_container
-        elsif service_pod.desired_state_unknown?
-          info "desired state is unknown for #{service_pod.name}, not doing anything"
-        elsif state_in_sync?(service_pod, service_container)
-          debug "state is in-sync: #{service_pod.desired_state}"
+        begin
+          ensure_desired_state
+        rescue => error
+          warn "failed to sync #{service_pod.name}: #{error}"
+          sync_state_to_master(current_state, error)
         else
-          warn "unknown state #{service_pod.desired_state} for #{service_pod.name}"
+          sync_state_to_master(current_state)
         end
+
+        self.terminate if service_pod.terminated?
```

Comment:
The `terminate` term is from celluloid, so we would have to rename the `service_pods.desired_state` :)

BTW what happens here now is that the server destroys the `GridServiceInstance`, and the agent signals `ServicePodWorker.destroy`... then if the `docker rm` fails, then the `ServicePodWorker` just logs a warning, and terminates itself. The docker container remains running, and the `ServicePodManager` doesn't pick that up until you restart the agent.

I'm not sure the existing implementation behaves better either... it will crash the `ServicePodWorker`, and then the `ServicePodManager` will not restart it.